stan/math/prim/prob/hmm_marginal_lpdf.hpp

```diff
@@ -0,0 +1,197 @@
+  const int n_transitions = omegas.cols() - 1;
+  alphas.col(0) = omegas.col(0).cwiseProduct(rho_val);
+
+  const auto norm = alphas.col(0).maxCoeff();
+  alphas.col(0) /= norm;
+  alpha_log_norms(0) = log(norm);
+
+  auto Gamma_val_transpose = Gamma_val.transpose().eval();
+  for (int n = 1; n <= n_transitions; ++n) {
+    alphas.col(n)
+        = omegas.col(n).cwiseProduct(Gamma_val_transpose * alphas.col(n - 1));
+    const auto col_norm = alphas.col(n).maxCoeff();
+    alphas.col(n) /= col_norm;
+    alpha_log_norms(n) = log(col_norm) + alpha_log_norms(n - 1);
+  }
+  norm_norm = alpha_log_norms(n_transitions);
+  return log(alphas.col(n_transitions).sum()) + norm_norm;
+}
+
+/**
+ * For a Hidden Markov Model with observation y, hidden state x,
+ * and parameters theta, return the log marginal density, log
+ * p(y | theta). In this setting, the hidden states are discrete
+ * and take values over the finite space {1, ..., K}.
+ * The marginal lpdf is obtained via a forward pass, and
+ * the derivative is calculated with an adjoint method,
+ * e.g (Betancourt, Margossian, & Leos-Barajas, 2020).
+ * log_omegas is a matrix of observational densities, where
+ * the (i, j)th entry corresponds to the density of the ith observation, y_i,
+ * given x_i = j.
+ * The transition matrix Gamma is such that the (i, j)th entry is the
+ * probability that x_n = j given x_{n - 1} = i. The rows of Gamma are
+ * simplexes.
+ * The Gamma argument is only checked if there is at least one transition.
+ *
+ * @tparam T_omega type of the log likelihood matrix
+ * @tparam T_Gamma type of the transition matrix
+ * @tparam T_rho type of the initial guess vector
+ * @param[in] log_omegas log matrix of observational densities.
+ * @param[in] Gamma transition density between hidden states.
+ * @param[in] rho initial state
+ * @return log marginal density.
+ * @throw `std::invalid_argument` if Gamma is not square, when we have
+ *         at least one transition, or if the size of rho is not the
+ *         number of rows of log_omegas.
+ * @throw `std::domain_error` if rho is not a simplex and of the rows
+ *         of Gamma are not a simplex (when there is at least one transition).
+ */
+template <typename T_omega, typename T_Gamma, typename T_rho>
+inline auto hmm_marginal_lpdf(
+    const Eigen::Matrix<T_omega, Eigen::Dynamic, Eigen::Dynamic>& log_omegas,
+    const Eigen::Matrix<T_Gamma, Eigen::Dynamic, Eigen::Dynamic>& Gamma,
+    const Eigen::Matrix<T_rho, Eigen::Dynamic, 1>& rho) {
+  using T_partial_type = partials_return_t<T_omega, T_Gamma, T_rho>;
+  using eig_matrix_partial
+      = Eigen::Matrix<T_partial_type, Eigen::Dynamic, Eigen::Dynamic>;
+  using eig_vector_partial = Eigen::Matrix<T_partial_type, Eigen::Dynamic, 1>;
+  int n_states = log_omegas.rows();
+  int n_transitions = log_omegas.cols() - 1;
+
+  check_consistent_size("hmm_marginal_lpdf", "rho", rho, n_states);
+  check_simplex("hmm_marginal_lpdf", "rho", rho);
+  if (n_transitions != 0) {
```

Comment:
I missed this conditional in the first review.  Semantically, you don't want special conditions to bypass the tests.  

It's also likely to slow the code down (not noticeably given everything else going on!) because the condition's almost never going to be satisfied in any realistic data set.

Reply:
The reason for having the conditional is  to allow the user to not construct Gamma if there are 0 transitions (as in it suffices to declare Gamma and make it 0 x 0). So I'd prefer keeping the conditional. But, no strong preference, it's your call.

Reply:
Presumably we don't need size 0 inputs at all.  For efficiency, those can just be removed from the data.  

How often do size 1 inputs come up?  I suspect that it'd be where the size 1 input is just one of many inputs, so `Gamma` would already exist.  This would save the check for simplex condition, which involves some real arithmetic.

In our other distributions, we've put the check for size 0 after all the checks.  For example, in `normal(y | mu, sigma)`, there's a check that if any of the arguments is a container (vector, row vector or array), then all of the container arguments have to be the same size.  If any of the inputs is size 0, we can return 0, but we do the size equality check and the check that `sigma > 0` first.

Reply:
I removed the conditional.

Now, a short comment.

> Presumably we don't need size 0 inputs at all. For efficiency, those can just be removed from the data.

The function always requires you to pass Gamma but if you have 0 transitions, you wouldn't need Gamma. I get that it's an unlikely occurrence but it's an edge case that we wanted to test and it's been its own Pandora box. Not sure about size 1, but even then, you want the element in Gamma to be 1.

Reply:
Thanks.  

I understand that it's not free to check that Gamma's a proper stochastic matrix and that it's not going to be used in the likelihood if there aren't any transitions.  Similarly, the outputs and initial distribution won't get used if the input is size 0.  Nevertheless, I'd like to keep the arguments legal everywhere just because the interface is simpler to doc and explain that way.  If the legal argument checks didn't apply for size 0 and/or 1 inputs, all the doc would need to reflect that, so every `@throw`and description of matrix would have to change.  So it'd be something like `@Gamma a stochastic matrix that's multiplicable by ...` with `@param Gamma a stochastic matrix that's multiplicable by ..., or if the size of the argument y is 0 or 1, it can be any size or shape matrix` and so on.